### PR TITLE
Keeping branches updated

### PR DESCRIPTION
**Pull request by lgh2**
Adds instructions to docs on keeping remote branches updated.

### CODE REVIEW

File: docs/development/submitting-patches.rst [hunk 1]
```diff
@@ -57,6 +57,61 @@ Important information should be documented with prose in the ``docs`` section.
 To ensure it builds and passes `doc8`_ style checks you can run
 ``make docs`` and ``make lint`` respectively.
 
+Keeping your local branch updated
+---------------------------------
+
+As you work, it is helpful to keep your local branch up-to-date with the PRs
+that are merged into the main Warehouse repository. Most people working on
+the project use this workflow.
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: The wording here implies that you're keeping a single branch up-to-date with multiple branches (since each PR is a branch). I suggest you revise to say something about keeping your local *master* branch up-to-date with the master branch in the main Warehouse repository, which moves forward as maintainers merge pull requests.

File: docs/development/submitting-patches.rst [hunk 2]
```diff
@@ -57,6 +57,61 @@ Important information should be documented with prose in the ``docs`` section.
...
+.. code-block:: console
+
+  $ git remote -v
+  origin  https://github.com/USERNAME/warehouse.git (fetch)
+  origin  https://github.com/USERNAME/warehouse.git (push)
+  upstream  https://github.com/pypa/warehouse.git (fetch)
+  upstream  https://github.com/pypa/warehouse.git (push)
+
+
+In the example above, USERNAME is your username on GitHub.
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: Please put backticks or similar around code-type words like USERNAME in this document.

File: docs/development/submitting-patches.rst [hunk 3]
```diff
@@ -57,6 +57,61 @@ Important information should be documented with prose in the ``docs`` section.
...
+  $ git fetch upstream
+
+Then, check out your master branch, and rebase the changes on top of it:
+
+.. code-block:: console
+
+  $ git checkout master
+  $ git rebase upstream/master
+
+Next, push the changes you have fetched to your origin branch on GitHub:
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: You mean your master branch on your origin repo?

File: docs/development/submitting-patches.rst [hunk 4]
```diff
@@ -57,6 +57,61 @@ Important information should be documented with prose in the ``docs`` section.
...
+
+.. code-block:: console
+
+  $ git checkout master
+  $ git push origin master
+
+Now your local branch and your origin are up-to-date with the most recent
+changes on the upstream Warehouse branch.
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: Please rewrite to distinguish branches and repositories more clearly.

File: docs/development/submitting-patches.rst [hunk 5]
```diff
@@ -57,6 +57,67 @@ Important information should be documented with prose in the ``docs`` section.
...
+  git fetch upstream
+
+Then, check out your master branch, and rebase the changes on top of it:
+
+.. code-block:: console
+
+  git checkout master
+  git rebase upstream/master
+
+Next, push the changes you have fetched to your ``origin`` branch on GitHub:
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: I'd reword as:

> Next, push the changes you have fetched to the master branch on your ``origin`` repository on GitHub:

File: docs/development/submitting-patches.rst (hunk 1 above)
Comment (brainwane) [reply docs/development/submitting-patches.rst]: I figure you're still working on [this](https://github.com/pypa/warehouse/pull/2760#discussion_r160774503) - just a reminder.

File: docs/development/submitting-patches.rst [hunk 6]
```diff
@@ -57,6 +57,67 @@ Important information should be documented with prose in the ``docs`` section.
...
+
+.. code-block:: console
+
+  git checkout master
+  git push origin master
+
+Now your local branch and your ``origin`` are up-to-date with the most recent
+changes on the ``upstream`` Warehouse branch.
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: I think you should use the word "master" more often here when referring to a master branch, to distinguish from remotes like "origin" & "upstream".

File: docs/development/submitting-patches.rst [hunk 7]
```diff
@@ -57,6 +57,67 @@ Important information should be documented with prose in the ``docs`` section.
...
+To keep your feature branches updated, the process is similar:
+
+.. code-block:: console
+
+   git checkout awesome-feature
+   git fetch upstream
+   git rebase upstream/master
+
+Now your feature branch has been updated with the latest changes from the
+``upstream`` Warehouse repository.
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: I'd reword as:

> Now your feature branch has been updated with the latest changes from the master branch on the ``upstream`` Warehouse repository.

File: docs/development/submitting-patches.rst [hunk 8]
```diff
@@ -57,6 +57,72 @@ Important information should be documented with prose in the ``docs`` section.
...
+repo have been updated with the most recent changes from the main Warehouse
+repository.
+
+To keep your feature branches updated, the process is similar:
+
+.. code-block:: console
+
+   git checkout awesome-feature
+   git fetch upstream
+   git rebase upstream/master
```
Comment (di) [inline docs/development/submitting-patches.rst]: Might be worth mentioning the next step here as well: force-pushing the branch to the remote.

Comment (lgh2) [reply docs/development/submitting-patches.rst]: Do you mean force-pushing the feature branch to ``origin`` while you are working?  Or for making a pull request, or both?

Comment (di) [reply docs/development/submitting-patches.rst]: Just to keep your branch up-to-date on your remote, whether you've already made a PR or not.

The `master` branch here moves somewhat quickly due to all the dependency upgrades that get merged, so it's pretty likely that a new contributor will have to force-push at least once while working on their pull request.

File: docs/development/submitting-patches.rst [hunk 9]
```diff
@@ -57,6 +57,112 @@ Important information should be documented with prose in the ``docs`` section.
...
+Once you have pushed your feature branch to your ``origin``, if you need to
+update it again, you will have to force push your changes by running the
+following command:
+
+.. code-block:: console
+
+    git push -f origin awesome-feature
+
+The ``-f`` flag after ``push`` updates your ``origin`` feature branch with the
+changes you have made on your local feature branch.
```
Comment (brainwane) [inline docs/development/submitting-patches.rst]: You should explain here that `-f` means "force" -- you are implying that it means "feature"!

You should also explain here that if you DO have a pull request open, updating the branch on `origin` will update the PR.

**Comment by lgh2** [comment]
@di Did you have any other changes you'd like me to make?
